File: python/src/workshop_api/fitness/managing_memberships/pause_membership/pause_membership_controller.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from workshop_api.fitness.database import get_db_session
from workshop_api.fitness.managing_memberships.pause_membership.pause_membership_request import (
    PauseMembershipRequest,
)
from workshop_api.fitness.managing_memberships.pause_membership.pause_membership_response import (
    PauseMembershipResponse,
)
from workshop_api.fitness.managing_memberships.shared.membership_entity import MembershipOrmModel

router = APIRouter(prefix="/api/memberships", tags=["membership"])


@router.post(
    "/{membership_id}/pause",
    response_model=PauseMembershipResponse,
    response_model_by_alias=True,
)
def pause_membership(
    membership_id: UUID,
    request: PauseMembershipRequest,
    session: Session = Depends(get_db_session),
) -> PauseMembershipResponse:
    if request.pause_start_date is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="pauseStartDate is required",
        )
    if request.pause_end_date is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="pauseEndDate is required",
        )
    if request.pause_end_date < request.pause_start_date:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="pauseEndDate must not be before pauseStartDate",
        )

    membership = session.get(MembershipOrmModel, str(membership_id))
    if membership is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Membership {membership_id} was not found",
        )
    if not membership.is_active():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Only active memberships can be paused",
        )
    if membership.end_date < date.today():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Expired memberships cannot be paused",
        )

    previous_status = membership.status
    previous_end_date = membership.end_date
    membership.pause(request.pause_start_date, request.pause_end_date, request.reason)
    session.commit()
    session.refresh(membership)

    return PauseMembershipResponse(
        membershipId=membership.id,
        previousStatus=previous_status,
        newStatus=membership.status,
        pauseStartDate=membership.pause_start_date,
        pauseEndDate=membership.pause_end_date,
        previousEndDate=previous_end_date,
        newEndDate=membership.end_date,
        reason=membership.pause_reason,
        message="Membership paused",
    )

File: java/src/main/java/com/workshop/architecture/fitness/pause_membership/PauseMembershipController.java
package com.workshop.architecture.fitness.pause_membership;

import com.workshop.architecture.fitness.shared.MembershipEntity;
import com.workshop.architecture.fitness.shared.MembershipRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.util.UUID;

@RestController
@RequestMapping("/api/memberships")
public class PauseMembershipController {

    private final MembershipRepository membershipRepository;

    public PauseMembershipController(MembershipRepository membershipRepository) {
        this.membershipRepository = membershipRepository;
    }

    @PostMapping("/{membershipId}/pause")
    PauseMembershipResponse pauseMembership(
            @PathVariable String membershipId,
            @RequestBody PauseMembershipRequest request
    ) {
        MembershipEntity membership;
        String previousStatus;
        LocalDate previousEndDate;

        if (request.pauseStartDate() == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "pauseStartDate is required");
        }

        if (request.pauseEndDate() == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "pauseEndDate is required");
        }

        if (request.pauseEndDate().isBefore(request.pauseStartDate())) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST,
                    "pauseEndDate must not be before pauseStartDate"
            );
        }

        membership = membershipRepository.findById(UUID.fromString(membershipId))
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Membership %s was not found".formatted(membershipId)
                ));

        if (!membership.isActive()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Only active memberships can be paused");
        }

        if (membership.getEndDate().isBefore(LocalDate.now())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Expired memberships cannot be paused");
        }

        previousStatus = membership.getStatus();
        previousEndDate = membership.getEndDate();

        membership.pause(request.pauseStartDate(), request.pauseEndDate(), request.reason());
        membership = membershipRepository.save(membership);

        return new PauseMembershipResponse(
                membership.getId().toString(),
                previousStatus,
                membership.getStatus(),
                membership.getPauseStartDate(),
                membership.getPauseEndDate(),
                previousEndDate,
                membership.getEndDate(),
                membership.getPauseReason(),
                "Membership paused"
        );
    }

}
